File: api/create-order.py
import os
import json
import urllib.request
import base64
from http.server import BaseHTTPRequestHandler

class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

    def do_POST(self):
        # Hosted Payment Links Canonical Mapping (No Orders API credentials required)
        hosted_links = {
            'QUANT_AUDIT_49': os.environ.get('PAYPAL_LINK_49', 'https://www.paypal.com/ncp/payment/SH9CKB2WSX728'),
            'QUANT_EXECUTION_REALITY_AUDIT_79': os.environ.get('PAYPAL_LINK_79', 'https://www.paypal.com/ncp/payment/TMMGL3YRC8PFN'),
            'COMPLETE_QUANT_VALIDATION_BUNDLE_96': os.environ.get('PAYPAL_LINK_96', 'https://www.paypal.com/ncp/payment/2Y3RX97HNWXY6')
        }
        
        content_length = int(self.headers.get('Content-Length', 0))
        body = {}
        if content_length > 0:
            try:
                body = json.loads(self.rfile.read(content_length).decode('utf-8'))
            except Exception:
                body = {}

        product_id = body.get('product_id', 'QUANT_AUDIT_49')
        payment_link = hosted_links.get(product_id, hosted_links['QUANT_AUDIT_49'])

        self.send_response(200)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(json.dumps({
            'status': 'DEPRECATED_MIGRATED_TO_HOSTED_LINKS',
            'product_id': product_id,
            'approvalUrl': payment_link,
            'message': 'PayPal Orders API has been migrated to PayPal Hosted Payment Links. Use direct hosted link.'
        }).encode())

File: api/create-order.js
const https = require('https');

function postRequest(url, headers, body) {
    return new Promise((resolve, reject) => {
        const req = https.request(url, { method: 'POST', headers }, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => {
                try {
                    resolve({ status: res.statusCode, data: JSON.parse(data) });
                } catch (e) {
                    resolve({ status: res.statusCode, data });
                }
            });
        });
        req.on('error', reject);
        if (body) req.write(body);
        req.end();
    });
}

module.exports = async (req, res) => {
    // Set CORS headers
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

    if (req.method === 'OPTIONS') {
        return res.status(200).end();
    }

    const clientId = process.env.PAYPAL_CLIENT_ID;
    const clientSecret = process.env.PAYPAL_CLIENT_SECRET;

    if (!clientId || !clientSecret) {
        return res.status(500).json({ error: 'Server PayPal credentials missing' });
    }

    try {
        // 1. Get OAuth Token
        const auth = Buffer.from(`${clientId}:${clientSecret}`).toString('base64');
        const tokenRes = await postRequest(
            'https://api-m.paypal.com/v1/oauth2/token',
            {
                'Authorization': `Basic ${auth}`,
                'Content-Type': 'application/x-www-form-urlencoded'
            },
            'grant_type=client_credentials'
        );

        if (tokenRes.status !== 200 || !tokenRes.data.access_token) {
            return res.status(500).json({ error: 'OAuth authentication failed', details: tokenRes.data });
        }

        const accessToken = tokenRes.data.access_token;

        const productId = (req.body && req.body.product_id) ? req.body.product_id : 'QUANT_AUDIT';
        const amountVal = (productId === 'QUANT_EXECUTION_REALITY_AUDIT') ? '79.00' : '49.00';
        const descVal = (productId === 'QUANT_EXECUTION_REALITY_AUDIT')
            ? 'Automaton Quant Execution Reality Audit ($79 USD)'
            : 'Automaton Quant Audit Verification ($49 USD)';

        // 2. Create Order
        const orderPayload = JSON.stringify({
            intent: 'CAPTURE',
            purchase_units: [{
                amount: {
                    currency_code: 'USD',
                    value: amountVal
                },
                description: descVal
            }],
            application_context: {
                brand_name: 'Automaton Quant Audit',
                landing_page: 'NO_PREFERENCE',
                user_action: 'PAY_NOW',
                return_url: 'https://jarmx75.github.io/Gemini-3-AlphaQuant-HUB-V1/sample.html?status=success',
                cancel_url: 'https://jarmx75.github.io/Gemini-3-AlphaQuant-HUB-V1/?status=cancelled'
            }
        });

        const orderRes = await postRequest(
            'https://api-m.paypal.com/v2/checkout/orders',
            {
                'Authorization': `Bearer ${accessToken}`,
                'Content-Type': 'application/json'
            },
            orderPayload
        );

        if (orderRes.status === 201 || orderRes.status === 200) {
            const approveLink = orderRes.data.links.find(l => l.rel === 'approve' || l.rel === 'payer-action');
            return res.status(200).json({
                orderID: orderRes.data.id,
                status: orderRes.data.status,
                approvalUrl: approveLink ? approveLink.href : `https://www.paypal.com/checkoutnow?token=${orderRes.data.id}`
            });
        } else {
            return res.status(500).json({ error: 'Failed to create PayPal order', details: orderRes.data });
        }

    } catch (err) {
        return res.status(500).json({ error: err.message });
    }
};
